pkg/traceql/engine_metrics.go

```diff
@@ -0,0 +1,531 @@
+	// Moving attributes to the first pass is like asserting non-nil on them.
+	// TODO
+
+	if dedupeSpans {
+		// We dedupe based on trace ID and start time.  Obviously this doesn't
+		// work if 2 spans have the same start time, but this doesn't impose any
+		// more data on the query when sharding is already present above (most cases)
+		traceID := NewIntrinsic(IntrinsicTraceID)
+		if !storageReq.HasAttribute(traceID) {
+			storageReq.Conditions = append(storageReq.Conditions, Condition{Attribute: traceID})
+		}
+	}
+
+	// (1) any overlapping trace
+	// TODO - Make this dynamic since it can be faster to skip
+	// the trace-level timestamp check when all or most of the traces
+	// overlap the window.
+	// storageReq.StartTimeUnixNanos = req.Start
+	// storageReq.EndTimeUnixNanos = req.End // Should this be exclusive?
+	// (2) Only include spans that started in this time frame.
+	//     This is checked inside the evaluator
+	me.checkTime = true
+	me.start = req.Start
+	me.end = req.End
+
+	// Avoid a second pass when not needed for much better performance.
+	// TODO - Is there any case where eval() needs to be called but AllConditions=true?
+	if !storageReq.AllConditions || len(storageReq.SecondPassConditions) > 0 {
+		storageReq.SecondPass = func(s *Spanset) ([]*Spanset, error) {
+			return eval([]*Spanset{s})
+		}
+	}
+
+	return me, nil
+}
+
+func lookup(needles []Attribute, haystack Span) Static {
+	for _, n := range needles {
+		if v, ok := haystack.AttributeFor(n); ok {
+			return v
+		}
+	}
+
+	return Static{}
+}
+
+type MetricsEvalulator struct {
+	start, end      uint64
+	checkTime       bool
+	dedupeSpans     bool
+	deduper         *SpanDeduper2
+	storageReq      *FetchSpansRequest
+	metricsPipeline metricsFirstStageElement
+	count           int
+	deduped         int
+	mtx             sync.Mutex
+}
+
+func (e *MetricsEvalulator) Do(ctx context.Context, f SpansetFetcher) error {
+	fetch, err := f.Fetch(ctx, *e.storageReq)
+	if errors.Is(err, util.ErrUnsupported) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if e.dedupeSpans && e.deduper == nil {
+		e.deduper = NewSpanDeduper2()
+	}
+
+	defer fetch.Results.Close()
+
+	for {
+		ss, err := fetch.Results.Next(ctx)
+		if err != nil {
+			return err
+		}
+		if ss == nil {
+			break
+		}
+
+		for _, s := range ss.Spans {
+			if e.checkTime {
+				st := s.StartTimeUnixNanos()
+				if st < e.start || st >= e.end {
+					continue
+				}
+			}
+
+			e.mtx.Lock()
+			if e.dedupeSpans && e.deduper.Skip(ss.TraceID, s.StartTimeUnixNanos()) {
+				e.deduped++
+				e.mtx.Unlock()
+				continue
+			}
+
+			e.count++
+			e.metricsPipeline.observe(s)
+			e.mtx.Unlock()
+		}
+
+		ss.Release()
+	}
+
+	return nil
+}
+
+func (e *MetricsEvalulator) SpanCount() {
+	fmt.Println(e.count, e.deduped)
+}
+
+func (e *MetricsEvalulator) Results() (SeriesSet, error) {
+	return e.metricsPipeline.result(), nil
+}
+
+// SpanDeduper2 is EXTREMELY LAZY. It attempts to dedupe spans for metrics
```

Comment:
This will be an area where we need to continue experimenting.  For I've chosen the approach that I think offers the best trade-offs. We can easily increase correctness at the cost of performance.

For now it dedupes using trace ID and span start time.  The downside is that it's obviously not 100% unique, but the performance is great because it doesn't require the query to fetch any additional columns, we already load them for backend queries (trace ID will used for sharding).   Deduping on trace ID/span ID is the obviously 100% correct method and where we started, but the span ID column is too hefty and significantly reduces performance.  A middle-ground option would be to use the nested set span ID (integer) which is also unique but lighter-weight than span ID. However the downsides are that it's only populated for complete traces, so we'd need some fallback for partial traces (back to using start time?), and there's no good way to access it directly. We decided not to expose it through the Fetch layer.

Additionally the data structure used is important too.  This current implementation is a bunch of maps of 32-bit hashes.  It has good performance but increased chance of collision vs 64-bit hashes (or storing traceID span ID in the maps directly).  We use the last byte of the trace ID to perform a single layer of 1-byte sharding which reduces the pressure on any single map.